tinygraph-poppy.c

```diff
@@ -0,0 +1,138 @@
+#include <math.h>
+#include <string.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "tinygraph-bits.h"
+#include "tinygraph-align.h"
+#include "tinygraph-poppy.h"
+
+
+typedef struct tinygraph_poppy_rank {
+  uint64_t *level0;
+  uint64_t level0_len;
+
+  tinygraph_bitset_s bitset_view;
+} tinygraph_poppy_rank;
+
+
+tinygraph_poppy_rank* tinygraph_poppy_rank_construct(tinygraph_bitset_s bitset) {
+  TINYGRAPH_ASSERT(bitset);
+
+  tinygraph_poppy_rank *out = malloc(sizeof(tinygraph_poppy_rank));
+
+  if (!out) {
+    return NULL;
+  }
+
+  // It's important that we align the index structures at
+  // the cache line boundary, which is 64 bytes nowadays
+  const uint32_t cacheline = 64;
+
+  const uint64_t *bits = tinygraph_bitset_get_data(bitset);
+  const uint64_t nbits = tinygraph_bitset_get_size(bitset);
+
+  // 512 bits basic blocks for now
+  const uint64_t level0_len = nbits / 512;
```

Comment:
Todo: we should add

    ceil(nbits / 512)

basic blocks, with the last one storing the overall cumsum for the total bitvector